app/models/notification.rb

```diff
@@ -138,17 +140,40 @@ def browserable(types: [], exclude_types: [], from_account_id: nil, include_filt
       end
     end
 
-    def paginate_groups(limit, pagination_order)
+    def paginate_groups(limit, pagination_order, grouped_types: nil)
       raise ArgumentError unless %i(asc desc).include?(pagination_order)
 
       query = reorder(id: pagination_order)
 
+      # Ideally `:types` would be a bind rather than part of the SQL itself, but that does not
+      # seem to be possible to do with Rails, considering that the expression would occur in
+      # multiple places, including in a `select`
+      group_key_sql = begin
+        if grouped_types.present?
+          # Normalize `grouped_types` so the number of different SQL query shapes remains small, and
+          # the queries can be analyzed in monitoring/telemetry tools
+          grouped_types = (grouped_types.map(&:to_sym) & GROUPABLE_NOTIFICATION_TYPES).sort
+
+          sanitize_sql_array([<<~SQL.squish, { types: grouped_types }])
+            COALESCE(
+              CASE
+                WHEN notifications.type IN (:types) THEN notifications.group_key
+                ELSE NULL
+              END,
+              'ungrouped-' || notifications.id
+            )
+          SQL
+        else
+          "COALESCE(notifications.group_key, 'ungrouped-' || notifications.id)"
+        end
+      end
```

Comment:
Not sure how to say this. I understand too little of the query following this, so I might be totally wrong. But from afar it looks as if when no `grouped_types` are passed in, it just reverts to the old query. That would still then group notifications, would it not? This might be surprising.

What *should* happen when someone calls the API without passing in `grouped_types`? Should this maybe disable grouping altogether? Might make some people happy.

Reply:
The whole API is geared towards having grouped notifications, if one does not want grouped notifications, `/api/v1/notifications`, for which we have no deprecation plan yet, is better suited (simpler API and underlying queries), so my idea was by default, you support all notification types to be grouped, which results in a simpler query. This would be the case of our web interface at all times.

Reply:
I tend to forget that the old API is going to stay around. Still, the behavior seemed surprising to me, so we should take care to document this properly.

And we should probably add a note that app developers worried about breaking changes should _always_ specificy all `grouped_types` they can handle.